File: sidekick/src/models/daily_entry.py
from datetime import datetime, timezone
from pydantic import AwareDatetime, BaseModel, Field, field_validator
from utils.datetime_utils import ensure_utc


class Restitution(BaseModel):
    sleep_hours: float | None = None
    sleep_quality: float | None = Field(None, ge=1, le=5)
    hrv: int | None = None
    resting_hr: int | None = None
    readiness: float | None = Field(None, ge=1, le=5)
    comment: str | None = None


class RestitutionSync(BaseModel):
    """One day's objective restitution fields sourced live from Intervals.icu.

    Not persisted — used to fill gaps in a manually-entered Restitution. No
    equivalent for sleep_quality/readiness/comment: those are the athlete's
    own subjective ratings, which Intervals.icu doesn't provide.
    """

    date: str  # YYYY-MM-DD
    sleep_hours: float | None = None
    hrv: int | None = None
    resting_hr: int | None = None


class ActivityAssessment(BaseModel):
    activity_id: int
    activity_name: str
    rpe: int = Field(..., ge=1, le=10, description="Rate of Perceived Exertion (1-10)")
    notes: str | None = None
    tags: list[str] = Field(default_factory=list)


class DailyEntry(BaseModel):
    athlete_id: int
    date: str  # YYYY-MM-DD
    restitution: Restitution | None = None
    activity_assessments: list[ActivityAssessment] = Field(default_factory=list)
    created_at: AwareDatetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: AwareDatetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    @field_validator("created_at", "updated_at", mode="before")
    @classmethod
    def _utc(cls, v):
        return ensure_utc(v)


class DailyEntryRequest(BaseModel):
    date: str  # YYYY-MM-DD
    restitution: Restitution | None = None
    activity_assessments: list[ActivityAssessment] = Field(default_factory=list)

File: frontend/src/api/dailyEntry.ts
import { apiFetch } from "./client"

export interface Restitution {
  sleep_hours?: number
  sleep_quality?: number
  hrv?: number
  resting_hr?: number
  readiness?: number
  comment?: string
}

export interface ActivityAssessment {
  activity_id: number
  activity_name: string
  rpe: number
  notes?: string
}

export interface DailyEntryRequest {
  date: string
  restitution?: Restitution
  activity_assessments: ActivityAssessment[]
}

export interface DailyEntry extends DailyEntryRequest {
  athlete_id: number
  created_at: string
  updated_at: string
}

export async function fetchDailyEntry(
  athleteId: number,
  date: string
): Promise<DailyEntry | null> {
  return apiFetch<DailyEntry | null>(`/api/v1/athlete/${athleteId}/daily-entry?date=${date}`)
}

export function saveDailyEntry(
  athleteId: number,
  entry: DailyEntryRequest
): Promise<DailyEntry> {
  return apiFetch<DailyEntry>(`/api/v1/athlete/${athleteId}/daily-entry`, {
    method: "POST",
    body: JSON.stringify(entry),
  })
}
